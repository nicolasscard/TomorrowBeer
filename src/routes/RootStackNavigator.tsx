import React, { useContext } from 'react';
import { View } from 'react-native';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';

import { BeerListScreen, BeerDetailScreen } from '../screens';
import { ThemeContext } from '../context/themeContext/ThemeContext';

export type StackParams = {
  BeerList: undefined,
  BeerDetail: {id: number},
}

const Stack = createStackNavigator<StackParams>();

export const RootStackNavigator = () => {
  const { theme } = useContext( ThemeContext );

  return (
    <View style={{ flex: 1, backgroundColor: theme.colors.background }}> 
      <NavigationContainer theme={ theme }>
        <Stack.Navigator initialRouteName="BeerList">
          <Stack.Screen name="BeerList" options={{ headerShown: false }} component={ BeerListScreen } />
          <Stack.Screen name="BeerDetail" options={{ headerShown: false }} component={ BeerDetailScreen } />
        </Stack.Navigator>
      </NavigationContainer>
    </View>
  );
}
