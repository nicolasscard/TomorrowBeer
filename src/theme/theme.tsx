import { ThemeState } from '../interfaces/themeState';

export const lightTheme: ThemeState = {
    dark: false,
    colors: {
        primary: '#9F3A52',
        background: 'white',
        card: 'white',
        text: 'black',
        border: '#e6e6e6',
        notification: 'black',
    },
}

export const darkTheme: ThemeState = {
    dark: false,
    colors: {
        primary: '#9F3A52',
        background: 'white',
        card: 'white',
        text: 'black',
        border: '#e6e6e6',
        notification: 'black',
    },
}
