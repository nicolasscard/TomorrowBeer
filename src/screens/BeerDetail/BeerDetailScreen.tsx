import React, { useContext, useEffect } from 'react';
import { Dimensions, Text, TouchableOpacity, View, ActivityIndicator } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import { StackScreenProps } from '@react-navigation/stack';

import { usePunkApi } from '../../hooks/usePunkApi';
import { Hop, Malt } from '../../interfaces/punkApiResponse';
import { ThemeContext } from '../../context/themeContext/ThemeContext';
import { StackParams } from '../../routes/RootStackNavigator';
import { beerDetailStyles } from './styles';
import { FadeInImage } from '../../components/FadeInImage';
import { ScrollView } from 'react-native-gesture-handler';

const windowWidth = Dimensions.get('window').width;
interface Props extends StackScreenProps<StackParams, 'BeerDetail'>{};

const DetailScreen = ({ route, navigation }: Props) => {
  const { beers, isLoading, loadBeers } = usePunkApi();
  const { theme } = useContext( ThemeContext );
  const styles = beerDetailStyles(theme);

  useEffect(() => {
    loadBeers({ids: route.params.id});
  }, []);

  if (beers.length === 1) {
    const { name, tagline, description } = beers[0];
    const { ingredients: { malt, hops, yeast}, food_pairing, first_brewed, image_url } = beers[0];
    const { brewers_tips, contributed_by} = beers[0];

    return (
      <SafeAreaView style={{ flex: 1, backgroundColor: theme.colors.background, padding: 20 }}>
        <ScrollView style={{ flex: 1 }}>

            {/* back button */}
            <TouchableOpacity
                onPress={() => navigation.goBack()}
                style={styles.primaryButton}
                activeOpacity={ 0.6 }
            >
              <View style={{  }}>
                <Icon
                    name="chevron-left"
                    color="white"
                    size={ windowWidth * 0.09 }
                />
              </View>
            </TouchableOpacity>

            {/* headers */}
            <View style={styles.header}>
              <Text style={styles.title}>{name}</Text>
              <Text style={styles.subtitle}>{tagline}</Text>
              <Text style={styles.descriptionText}>{description}</Text>
            </View>

            {/* details */}
            <View style={{ flexDirection: 'row', marginBottom: 20, justifyContent: 'space-between' }}>
              <View style={styles.detailContent}>
                <Text style={styles.sectionTitle}>Ingredients:</Text>
                <Text>{`- Malts:`}</Text>
                <Text>{malt.map((value: Malt) => `   * ${value.name}\n`)}</Text>
                <Text>{`- Hops:`}</Text>
                <Text>{hops.map((value: Hop) => `   * ${value.name}\n`)}</Text>
                <Text>{`- Yeast:`}</Text>
                <Text>{yeast}</Text>
                <Text style={styles.sectionTitle}>First brewed</Text>
                <Text>{first_brewed}</Text>
                <Text style={styles.sectionTitle}>Is ideal to accompany with:</Text>
                <Text style={styles.descriptionText}>{food_pairing.map((value: string) => `* ${value}\n`)}</Text>
              </View>
              <View style={styles.image}>
                <FadeInImage
                  source={{ uri: image_url }}
                  style={ styles.beerImage }
                />
              </View>
            </View>

            {/* other information */}
            <View style={styles.bottomContent}>
              <Text style={styles.descriptionText}>{brewers_tips}</Text>
              <Text style={{ ...styles.descriptionText, textAlign: 'center'}}>{contributed_by}</Text>
            </View>

          </ScrollView>
      </SafeAreaView>
    );
  }
  else {
    <SafeAreaView style={{ flex: 1, backgroundColor: theme.colors.background, padding: 20 }}>
      <ActivityIndicator 
        style={{ flex: 1}} 
        color={ theme.colors.primary }
        size={ 30 }
      />
    </SafeAreaView>
  }

}

export default DetailScreen;
