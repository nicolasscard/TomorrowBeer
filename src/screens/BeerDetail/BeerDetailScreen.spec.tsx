import 'react-native-vector-icons/mock';
import React from 'react';
import renderer from 'react-test-renderer';
import { StackNavigationProp } from '@react-navigation/stack';
import { RouteProp } from '@react-navigation/native';

import { StackParams } from '../../routes/RootStackNavigator';
import { lightTheme } from '../../theme/theme';
import BeerDetailScreen from './BeerDetailScreen';

type NavigationProps = StackNavigationProp<StackParams, 'BeerDetail', undefined>;
type PartialNavigationProps = Partial<NavigationProps>;

const mockThemeContextValue = { 
  theme: lightTheme,
  setDarkTheme: jest.fn(),
  setLightTheme: jest.fn(),
};

jest.mock('react', () => {
  const React = jest.requireActual('react');
  return {
    ...React,
    useContext: jest.spyOn(React, 'useContext'),
  }
});

jest.mock('../../context/themeContext/ThemeContext', () => ({
  ThemeContext: {
    Consumer: ({ children }) => children(mockThemeContextValue),
  },
}));


describe('HomeScreen', () => {

    it('should render correctly', () => {
      jest.spyOn(React, 'useContext').mockReturnValue(mockThemeContextValue);

      const mockNavigate: PartialNavigationProps = {
        navigate: jest.fn(),
      };
  
      const mockRoute: RouteProp<StackParams, 'BeerDetail'> = {
        key: '123',
        name: 'BeerDetail',
        params: { id: 1 },
      };

      const component = renderer.create(
        <BeerDetailScreen navigation={mockNavigate as NavigationProps} route={mockRoute} /> 
      );
        
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  