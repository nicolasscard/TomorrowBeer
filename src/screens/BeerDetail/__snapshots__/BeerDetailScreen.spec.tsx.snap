// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeerDetail should render correctly 1`] = `null`;
